Implement matchMedia and MediaQueryList

metajack opened the pull request:
<!-- Please describe your changes on the following line: -->

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #13376 (github issue number if applicable).

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

Fixes #13376.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/13453)

<!-- Reviewable:end -->

metajack committed:
Implement matchMedia and MediaQueryList

Fixes #13376.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/lib.rs, components/style/media_queries.rs
 * @kichjang: components/script/dom/mod.rs, components/script/dom/webidls/MediaQueryList.webidl, components/script/dom/bindings/trace.rs, components/script/script_thread.rs, components/script/dom/window.rs, components/script/dom/webidls/Window.webidl, components/script/dom/mediaquerylist.rs

metajack commented:
I suspect I will still need to update some test expectations.

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit eea339b with merge fcc8645...

metajack commented:
@ms2ger @jdm I need signoff from one of you one MediaQueryList in the global according to the WPT failure.

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/3246)

jdm commented:
You have my blessing for exposing the new interface.

metajack commented:
@bors-servo try

Tests and expectations updated, and I changed the exposure to just the window object.

bors-servo commented:
:hourglass: Trying commit 037e14d with merge 30ff329...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/3248)

jdm commented:
If possible, try making Window only exposed to Window. Otherwise, re-expose MediaQueryList to Worker and update the appropriate test.

metajack commented:
@bors-servo try

@jdm Even though it isn't implemented, adding the annotation to matchMedia is enough to satisfy the WebIDL parser, so I did that.

bors-servo commented:
:hourglass: Trying commit c455bdf with merge cd6aaed...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/3249)

Wafflespeanut commented:
r? @jdm

metajack commented:
@bors-servo try

Fixed the unit tests and geckolib build.

bors-servo commented:
:hourglass: Trying commit ef4f28e with merge 2f2cb44...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/3448)

highfive commented:
▶ Unexpected subtest result in /cssom-view-1_dev/html/window-interface.htm:
      └ PASS [expected FAIL] window_functions

metajack commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit b422d97 with merge 6bc859a...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/3450)

metajack commented:
The two failures didn't happen in previous runs and seem to be new intermittents.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13323) made this pull request unmergeable. Please resolve the merge conflicts.

metajack commented:
@jdm review ping :)

jdm commented on components/style/lib.rs:
```diff
@@ -158,3 +160,16 @@ pub fn arc_ptr_eq<T: 'static>(a: &Arc<T>, b: &Arc<T>) -> bool {
+pub fn serialize_comma_separated_list<W, T>(dest: &mut W, list: &[T])
+                                            -> fmt::Result where W: fmt::Write, T: ToCss {
+    if list.len() == 0 {
+        write!(dest, "")
```
Is this necessary?

metajack replied:
Fixed.

jdm commented on components/script/dom/mediaquerylist.rs:
```diff
@@ -0,0 +1,80 @@
+        let viewport_size = TypedSize2D::from_untyped(&au_rect_to_f32_rect(current_viewport).size);
+        let device = Device::new(MediaType::Screen, viewport_size);
+        let result = self.media_query_list.evaluate(&device);
+        result
```
`let result = ...; result` can just be `...`;

metajack replied:
Fixed. It was leftover from some earlier debugging.

jdm commented on components/script/dom/mediaquerylist.rs:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    // https://drafts.csswg.org/cssom-view/#dom-mediaquerylist-onchange
+    event_handler!(change, GetOnchange, SetOnchange);
```
In general we try to avoid adding event handler attributes for events that we don't support yet.

metajack replied:
It's implemented now, so I left it.

jdm commented on tests/wpt/mozilla/tests/css/matchMedia.html:
```diff
@@ -0,0 +1,188 @@
+            }, "MediaQueryList.matches for \"(min-aspect-ratio: 1/1)\"");
+
+            test(function(){
+                mql = iframe_window.matchMedia("(width: 200px)");
```
This test doesn't pass right now; can we investigate why?

hbnrmx replied:
FWIW: Looks like this was fixed in #14192

jdm reviewed (changes requested):
I'm a bit leery of introducing an MQL implementation that doesn't do the live updates. I'd almost be tempted to pref it off by default rather than expose it in the current state.

metajack commented:
@jdm I couldn't find any similar code for dispatching events at the event loop turn. I tried to put that comment where I thought they should go. Do we do this somewhere else? I'm not opposed to adding it to this PR, but  I wasn't sure where exactly it needed to go.

metajack commented on tests/wpt/mozilla/tests/css/matchMedia.html:
```diff
@@ -0,0 +1,188 @@
+            }, "MediaQueryList.matches for \"(min-aspect-ratio: 1/1)\"");
+
+            test(function(){
+                mql = iframe_window.matchMedia("(width: 200px)");
```
It doesn't pass because we don't have support for that form https://github.com/servo/servo/blob/master/components/style/media_queries.rs#L28

jdm commented:
I believe the best way to handle this right now is to add it after the "resize" event in `handle_resize_event`, since that code is implementing 7.5. We'll need to store the last matching state of each MQL so we can determine if the matching state changed or not, too.

jdm commented:
My comments about using weak references got marked as outdated for some reason, but I still think it would be best to do it right from the start.

metajack commented on components/script/dom/window.rs:
```diff
@@ -277,6 +280,9 @@ pub struct Window {
     console_timers: TimerSet,
+
+    /// All the MediaQueryLists we need to update
+    media_query_lists: DOMRefCell<Vec<JS<MediaQueryList>>>,
```
We can't store WeakRefVec directly because it has not JSTraceable implementation (and seemingly can't have one). I did what Range does and made a WeakMediaQueryListVec type which wraps WeakVec.

metajack commented on components/script/dom/window.rs:
```diff
@@ -902,6 +912,17 @@ impl WindowMethods for Window {
+        let media_query_list = media_queries::parse_media_query_list(&mut parser);
+        let document = self.Document();
+        let mql = MediaQueryList::new(&self, &document, media_query_list);
+        // TODO: This should use WeakRefs so we don't hold on to every MQL forever
```
Done.

metajack commented on components/script/script_thread.rs:
```diff
@@ -791,6 +791,9 @@ impl ScriptThread {
             });
 
+            // https://html.spec.whatwg.org/multipage/#event-loop-processing-model
+            // TODO: Step 7.7 - evaluate media queries and report changes
```
I went ahead an implemented this.

metajack commented:
Ok, I fixed everything, and more tests now pass.

I may have regressed the `min-width: 150px` test, so I will investigate that, but otherwise this is ready for review again.

metajack commented:
Ok. I switched from current_viewport to window_size and that now gives real data instead of 0x0. It seems Window has nothing that knows the actual ViewportPx version of the size, which is unfortunate, but can be fixed when we add back zoom support some day.

Now `min-width: 150` also passes.

metajack commented:
@jdm re-r?

jdm commented on components/script/dom/window.rs:
```diff
@@ -234,6 +238,9 @@ pub struct Window {
     scroll_offsets: DOMRefCell<HashMap<UntrustedNodeAddress, Point2D<f32>>>,
+
+    /// All the MediaQueryLists we need to update
+    media_query_lists: DOMRefCell<WeakMediaQueryListVec>,
```
No need for both DOMRefCell and UnsafeCell.

jdm commented on components/script/dom/mediaquerylist.rs:
```diff
@@ -0,0 +1,166 @@
+        WeakMediaQueryListVec { cell: UnsafeCell::new(WeakRefVec::new()) }
+    }
+
+    pub fn push(&self, ref_: WeakRef<MediaQueryList>) {
```
Let's make this take a `&MediaQueryList` argument instead.

jdm reviewed (changes requested):
A couple small changes and this should be ready to go!

metajack commented:
@jdm I have addressed all your comments except for deriving JSTraceable, which isn't possible since WeakVec doesn't implement it.

Ready for re-review.

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 8c2b25b has been approved by `jdm`

<!-- @bors-servo r=jdm 8c2b25b8289fcf60001e8427255db71a937d8c67 -->

bors-servo commented:
:hourglass: Testing commit 8c2b25b with merge cd9b0b4...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/564)

KiChjang commented:
```
Tests with unexpected results:
  ▶ TIMEOUT [expected CRASH] /html/semantics/tabular-data/the-table-element/caption-methods.html
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  │ 3.3 (Core Profile) Mesa 12.0.1
  └ Insertion failed: HierarchyRequest (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }, at ../src/libcore/result.rs:799)

  ▶ Unexpected subtest result in /_mozilla/css/matchMedia.html:
  └ PASS [expected FAIL] MediaQueryList.matches for "(min-width: 150px)"
```

jdm commented:
Filed #14026 for the first failure, but the other needs to be updated.

metajack commented:
@bors-servo r=jdm

Updated test expectation.

bors-servo commented:
:pushpin: Commit 138a048 has been approved by `jdm`

<!-- @bors-servo r=jdm 138a0480fee8e6f0a338d85160aca49f14e245c9 -->

bors-servo commented:
:hourglass: Testing commit 138a048 with merge 6ef46ab...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3836), [arm64](http://build.servo.org/builders/arm64/builds/3846), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3924), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/565), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/566), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3914), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3906), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/595), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/602), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1828)
